### DISCUSSION TITLE
Package installation from PackageCloud results in 404 responses

### DISCUSSION THREAD

**Discussion by AphidGit** [rabbitmq/rabbitmq-website#1635] [General]
### Describe the bug

Recommended repository for installing, as per; 

https://www.rabbitmq.com/install-debian.html

includes as an option the 'packagecloud' repository. 

### Reproduction steps

1. Trying to configure apt with this repository fails to update, even after manually fixing the broken key scripts (just trying manually apt-key add and removing the signed-by requirement) and apt config file suggested by the site.

`404  Not Found [IP: 2600:9000:238d:7200:e:f4d2:20c0:93a1 443]`

2. navigate to the provided package cloud link, https://packagecloud.io/rabbitmq/rabbitmq-server/ubuntu/ results in 404. 

### Expected behavior

A quick check shows that https://packagecloud.io/rabbitmq/rabbitmq-server/ is accessible and hosts a bunch of rabbitmq versions and other folders.

I'm not sure where the instructions for debian/ubuntu should point to (but definitely not to a nonexistent path). Probably something like https://packagecloud.io/rabbitmq/rabbitmq-server/**packages**/ubuntu instead of https://packagecloud.io/rabbitmq/rabbitmq-server/ubuntu, please verify though.

_Side note: I personally hate certain companies for constantly changing their links and breaking everyone else's guides/instructions on how to do things like installing and configuring software. A much better solution is to change the dead links to redirects, so they don't break for the 10248 users that copied your instructions to their own sites, the internet is a broken enough mess as it is. Though perhaps this is the fault of packagecloud._

**Comment by michaelklishin** [comment]
We cannot do anything about stale pat data or IPs that may resolve but become unreachable.

**Comment by michaelklishin** [comment]
`https://packagecloud.io/rabbitmq/rabbitmq-server/ubuntu/` is used in an apt source list file. It is not meant to serve anything when requested from a Web browser; apt uses those repository URLs as roots, the actual files are nested deeper. apt uses both conventions as well as package metadata.

**Comment by michaelklishin** [comment]
PackageCloud has a way of setting up a repo that [can be found on the repository page]():

```
curl -s https://packagecloud.io/install/repositories/rabbitmq/rabbitmq-server/script.deb.sh | sudo bash
```

it is listed in our Debian installation guide.

The following [Quick Start script](https://rabbitmq.com/install-debian.html#apt-quick-start-packagecloud) works just fine in a new Ubuntu 20.04 container:

``` shell
#!/usr/bin/sh

sudo apt-get install curl gnupg apt-transport-https -y

## Team RabbitMQ's main signing key
curl -1sLf "https://keys.openpgp.org/vks/v1/by-fingerprint/0A9AF2115F4687BD29803A206B73A36E6026DFCA" | sudo gpg --dearmor | sudo tee /usr/share/keyrings/com.rabbitmq.team.gpg > /dev/null
## Launchpad PPA that provides modern Erlang releases
curl -1sLf "https://keyserver.ubuntu.com/pks/lookup?op=get&search=0xf77f1eda57ebb1cc" | sudo gpg --dearmor | sudo tee /usr/share/keyrings/net.launchpad.ppa.rabbitmq.erlang.gpg > /dev/null
## PackageCloud RabbitMQ repository
curl -1sLf "https://packagecloud.io/rabbitmq/rabbitmq-server/gpgkey" | sudo gpg --dearmor | sudo tee /usr/share/keyrings/io.packagecloud.rabbitmq.gpg > /dev/null

## Add apt repositories maintained by Team RabbitMQ
sudo tee /etc/apt/sources.list.d/rabbitmq.list <<EOF
## Provides modern Erlang/OTP releases
##
## "bionic" as distribution name should work for any reasonably recent Ubuntu or Debian release.
## See the release to distribution mapping table in RabbitMQ doc guides to learn more.
deb [signed-by=/usr/share/keyrings/net.launchpad.ppa.rabbitmq.erlang.gpg] http://ppa.launchpad.net/rabbitmq/rabbitmq-erlang/ubuntu bionic main
deb-src [signed-by=/usr/share/keyrings/net.launchpad.ppa.rabbitmq.erlang.gpg] http://ppa.launchpad.net/rabbitmq/rabbitmq-erlang/ubuntu bionic main

## Provides RabbitMQ
##
## "bionic" as distribution name should work for any reasonably recent Ubuntu or Debian release.
## See the release to distribution mapping table in RabbitMQ doc guides to learn more.
deb [signed-by=/usr/share/keyrings/io.packagecloud.rabbitmq.gpg] https://packagecloud.io/rabbitmq/rabbitmq-server/ubuntu/ bionic main
deb-src [signed-by=/usr/share/keyrings/io.packagecloud.rabbitmq.gpg] https://packagecloud.io/rabbitmq/rabbitmq-server/ubuntu/ bionic main
EOF

## Update package indices
sudo apt-get update -y

## Install Erlang packages
sudo apt-get install -y erlang-base \
                        erlang-asn1 erlang-crypto erlang-eldap erlang-ftp erlang-inets \
                        erlang-mnesia erlang-os-mon erlang-parsetools erlang-public-key \
                        erlang-runtime-tools erlang-snmp erlang-ssl \
                        erlang-syntax-tools erlang-tftp erlang-tools erlang-xmerl

## Install rabbitmq-server and its dependencies
sudo apt-get install rabbitmq-server -y --fix-missing
```

My best guess is that your OS needs an `apt update -y` run because it has old IP addresses (of PackageCloud or some other repository) cached by apt.

In addition, your understanding of how apt fetches metadata and packages is not correct: the URLs
you see in source list files are not meant to be browsed, they are repository roots that apt appends distribution and architecture names to.

**Comment by michaelklishin** [comment]
One thing that must be considered is that

```
curl -s https://packagecloud.io/install/repositories/rabbitmq/rabbitmq-server/script.deb.sh | sudo bash
```

only sets up the RabbitMQ repository. It does not set up a modern Erlang repo (such as our PPA or the Erlang repo on PackageCloud) but the Quick Start script above does both.

When updated to use `jammy` for distribution name instead of `bionic`, it works just fine on Ubuntu 22.04:

``` shell
#!/usr/bin/sh

sudo apt-get install curl gnupg apt-transport-https -y

## Team RabbitMQ's main signing key
curl -1sLf "https://keys.openpgp.org/vks/v1/by-fingerprint/0A9AF2115F4687BD29803A206B73A36E6026DFCA" | sudo gpg --dearmor | sudo tee /usr/share/keyrings/com.rabbitmq.team.gpg > /dev/null
## Launchpad PPA that provides modern Erlang releases
curl -1sLf "https://keyserver.ubuntu.com/pks/lookup?op=get&search=0xf77f1eda57ebb1cc" | sudo gpg --dearmor | sudo tee /usr/share/keyrings/net.launchpad.ppa.rabbitmq.erlang.gpg > /dev/null
## PackageCloud RabbitMQ repository
curl -1sLf "https://packagecloud.io/rabbitmq/rabbitmq-server/gpgkey" | sudo gpg --dearmor | sudo tee /usr/share/keyrings/io.packagecloud.rabbitmq.gpg > /dev/null

## Add apt repositories maintained by Team RabbitMQ
sudo tee /etc/apt/sources.list.d/rabbitmq.list <<EOF
## Provides modern Erlang/OTP releases
deb [signed-by=/usr/share/keyrings/net.launchpad.ppa.rabbitmq.erlang.gpg] http://ppa.launchpad.net/rabbitmq/rabbitmq-erlang/ubuntu jammy main
deb-src [signed-by=/usr/share/keyrings/net.launchpad.ppa.rabbitmq.erlang.gpg] http://ppa.launchpad.net/rabbitmq/rabbitmq-erlang/ubuntu jammy main

## Provides RabbitMQ
deb [signed-by=/usr/share/keyrings/io.packagecloud.rabbitmq.gpg] https://packagecloud.io/rabbitmq/rabbitmq-server/ubuntu/ jammy main
deb-src [signed-by=/usr/share/keyrings/io.packagecloud.rabbitmq.gpg] https://packagecloud.io/rabbitmq/rabbitmq-server/ubuntu/ jammy main
EOF

## Update package indices
sudo apt-get update -y

## Install Erlang packages
sudo apt-get install -y erlang-base \
                        erlang-asn1 erlang-crypto erlang-eldap erlang-ftp erlang-inets \
                        erlang-mnesia erlang-os-mon erlang-parsetools erlang-public-key \
                        erlang-runtime-tools erlang-snmp erlang-ssl \
                        erlang-syntax-tools erlang-tftp erlang-tools erlang-xmerl

## Install rabbitmq-server and its dependencies
sudo apt-get install rabbitmq-server -y --fix-missing
```

```
root@5e855e220ea3:/# service rabbitmq-server start
 * Starting message broker rabbitmq-server                                                                                                                                                                                                                                                         [ OK ]
root@5e855e220ea3:/# rabbitmq-diagnostics status
warning: the VM is running with native name encoding of latin1 which may cause Elixir to malfunction as it expects utf8. Please ensure your locale is set to UTF-8 (which can be verified by running "locale" in your shell)
Status of node rabbit@5e855e220ea3 ...
Runtime

OS PID: 5360
OS: Linux
Uptime (seconds): 11
Is under maintenance?: false
RabbitMQ version: 3.11.13
RabbitMQ release series support status: supported
Node name: rabbit@5e855e220ea3
```

**Comment by michaelklishin** [comment]
The Quick Start script [now performs an `apt update -y` first thing](https://github.com/rabbitmq/rabbitmq-website/commit/bdcb279379aaf162decd401a1fbbe8e7195e49e2), even before installing packages from standard repos, as their IP addresses can be cached just like those resolved for packagecloud.io.

Nothing else in the Debian installation guide seems off to me. The above script is functional, recommended first and broken down with a more detailed explanation later for those who would like to learn more.
